internal/oci/runtime_oci.go

```diff
@@ -898,73 +899,82 @@ func (r *runtimeOCI) AttachContainer(c *Container, inputStream io.Reader, output
 	return nil
 }
 
-// PortForwardContainer forwards the specified port provides statistics of a container.
-func (r *runtimeOCI) PortForwardContainer(c *Container, port int32, stream io.ReadWriter) error {
-	emptyStreamOnError := true
-	defer func() {
-		if emptyStreamOnError {
-			go func() {
-				if _, copyError := pools.Copy(ioutil.Discard, stream); copyError != nil {
-					logrus.Errorf("error closing port forward stream after other error: %v", copyError)
-				}
-			}()
+// PortForwardContainer forwards the specified port into the provided container.
+func (r *runtimeOCI) PortForwardContainer(ctx context.Context, c *Container, netNsPath string, port int32, stream io.ReadWriteCloser) error {
+	log.Infof(ctx,
+		"Starting port forward for %s in network namespace %s", c.ID(), netNsPath,
+	)
+
+	// Adapted reference implementation:
+	// https://github.com/containerd/cri/blob/8c366d/pkg/server/sandbox_portforward_unix.go#L65-L120
+	if err := ns.WithNetNSPath(netNsPath, func(_ ns.NetNS) error {
+		defer stream.Close()
+
+		conn, err := net.Dial("tcp", net.JoinHostPort("localhost", strconv.Itoa(int(port))))
```

Comment:
Oh, I think I see: I said you should do `"localhost"` to guarantee you get the same IP as the pod does, but this code isn't running fully inside the pod; it's in the pod's network namespace, but the _node_'s mount namespace, so it sees the node's `/etc/hosts`, `/etc/resolv.conf`, `/etc/nsswitch.conf`, so there's no reason to think that it's going to resolve "localhost" the same way as the pod would. Blah.

So ok, basically, trying to be dual-stack-ready is just a lose at this point. Hardcode to 127.0.0.1, and be ready to rewrite in some way when new upstream tests are added...

Reply:
yeah, that's why  `conn, err := net.Dial("tcp4", fmt.Sprintf("localhost:%d", port))` previously
https://github.com/cri-o/cri-o/pull/3749#discussion_r423737909

Reply:
Changed the implementation to match the latest comments. Big thanks again! :pray: